src/dbnode/storage/throttler.go

```diff
@@ -0,0 +1,142 @@
+package storage
+
+import (
+	"sync"
+)
+
+// Throttler controls fair access to limited resources.
+type Throttler struct {
+	sync.Mutex
+
+	keyState map[string]*keyContext
+	// Each entry in the queue should be unique to avoid unfair access
+	keyQueue []string
```

Comment:
nit: considering [`list.List`](https://golang.org/pkg/container/list/) for the queue so that you aren't constantly churning the slice during re-ordering.